tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -0,0 +1,321 @@
+  const auto loc = dominant_op->getLoc();
+  b.setInsertionPoint(root_ops.back());
+  auto zero = b.create<ConstantOp>(loc, b.getIndexType(),
+                                   b.getIntegerAttr(b.getIndexType(), 0));
+  auto one = b.create<ConstantOp>(loc, b.getIndexType(),
+                                  b.getIntegerAttr(b.getIndexType(), 1));
+  auto num_elements = codegen_utils::calcNumElements(b, loc, dominant_op);
+  Value var;
+  if (parallel_loop) {
+    SmallVector<Value, 2> vars;
+    (void)createParallelAndSetInsPt(b, loc, vars, ArrayRef<Value>{zero},
+                                    ArrayRef<Value>{num_elements},
+                                    ArrayRef<Value>{one}, ArrayRef<Value>{});
+    var = vars[0];
+  } else {
+    (void)createLoopAndSetInsPt(b, loc, var, zero, num_elements, one,
+                                ArrayRef<Value>{});
+  }
+  for (auto* root_op : root_ops) {
+    if (!succeeded(
+            lowerHelper(b, loc, root_op, var, shape_constraint_analysis))) {
+      return failure();
+    }
+  }
+  // remove the root_op if it has no other users except the memref
+  if (non_fusion) {
+    for (auto* root_op : root_ops) {
+      root_op->erase();
+    }
+  } else {
+    assert(parent != nullptr && "Parent must be provided for fusion lowering");
+    cleanUnusedLhloOps(parent);
+  }
+  b.create<scf::YieldOp>(loc, ValueRange({}));
+  return success();
+}
+
+bool isOnGpu(Operation* op) {
+  if (isa<FusionOp>(op)) {
+    // TODO: Revisit this when fusion on cpu is suppported
+    return true;
+  }
+  assert(isa<LmhloOp>(op) && "Unexpected usage of isOnGpu");
+  auto result_memref = cast<LmhloOp>(op).getResultBuffer();
+  auto memory_space =
+      result_memref.getType().cast<MemRefType>().getMemorySpace();
+  return memory_space && memory_space.isa<StringAttr>() &&
+         memory_space.cast<StringAttr>().getValue() == placement_utils::c_gpu;
+}
+
+}  // namespace
+
+// Expand the root ops in a fused func into a set of nested loops.
+// This pass must be executed after the fusion pass, and works together with
+// the InputInlineFusion pass after it for fusion codegen.
+struct LhloLegalizeRootsToParallelLoops
+    : public LhloLegalizeRootsToParallelLoopsPassBase<
+          LhloLegalizeRootsToParallelLoops> {
+  void runOnFunction() override {
+    auto func = getFunction();
+    OpBuilder b(func);
+    SmallVector<Operation*, 4> gpu_non_fusion_worklist;
+    SmallVector<Operation*, 4> cpu_non_fusion_worklist;
+    SmallVector<Operation*, 4> gpu_fusion_worklist;
+    func.walk([&](LmhloOp op) {
+      // Skip nested computation of reduce etc.
+      if (op->getParentOp() != func.getOperation()) {
+        return;
+      }
+      if (isa<FusionOp>(op)) {
+        // TODO: Revisit this when fusion on cpu is supported
+        gpu_fusion_worklist.push_back(op);
+      } else if (isOnGpu(op)) {
+        gpu_non_fusion_worklist.push_back(op);
```

Comment:
We didn't put single unfused nodes in a lmhlo.fusion, just feel it to be more intuitive, I think the additional complexity it brings for the codegen pass is acceptable.

It's another story for the lmhlo.FusionOp to have lmhlo inside, not mhlo. That mainly because we now do fusion on lmhlo, not mhlo. Refer to previous discussion here: https://groups.google.com/g/disc-mhlo/c/7z5Mitg4jVs
The fusion pass can be moved to mhlo after shape dialect is brought in to represent shape calculation on tensor layer and we would be able to do shape calculation lowering for mhlo.FusionOp. After these are done, a lmhlo.FusionOp with mhlo inside would make more sense, since its representation is more friendly to FusedIrEmitter.

BTW, since FusedIREmitter can only support pure static shape kernels. Do you think we might fully replace it someday?

Reply:
> It's another story for the lmhlo.FusionOp to have lmhlo inside, not mhlo. That mainly because we now do fusion on lmhlo, not mhlo. Refer to previous discussion here: https://groups.google.com/g/disc-mhlo/c/7z5Mitg4jVs

Even if you do the fusion on lmhlo, you could promote the body to operate on mhlo when performing the fusion?

Reply:
The mhlo ops has already been converted to lmhlo ops before the fusion pass. Thus the body of the lmhlo.FusionOp will be made of lmhlo ops. I'm not sure what do you mean by 'promote', could you explain a little bit more?

Reply:
by promotion, I meant actually have the lhlo fusion consuming/producing the buffers, but its body operates on tensor.
Anyway this is just transient as you mentioned above, when we'll migrate to a fusion on at the tensor level.

Reply:
I'm aware of the mhlo-disc group discussion, but since it's a private list, I was more of asking publishing the rationale in the source code as comments, or as an individual, public .md document.

Reply:
I add some comments to interpret the rationale of lmhlo.fusion representation.